samples: convert smart amp test module to not expect the blob during init with IPC4

ujfalusi opened the pull request:
Hi,

the module initialization for IPC4 does not contain binary (configuration) blob, only base_cfg and optional bae_cfg_ext.
The configuration blob is sent as a large_config message separately.

lgirdwood commented:
@wszypelt @lrudyX good to go here ? Dont think this flow is tested in CI yet ?

wszypelt commented:
@lgirdwood 3 tests on KD was failed, I think it's not good to merge

ujfalusi commented:
> @lgirdwood 3 tests on KD was failed, I think it's not good to merge

To be fair, most of the PRs are failing in a same way, if I read the logs right, no?

wszypelt commented:
@ujfalusi a lot of PR has a problem with one test and right now I'm trying to understand why, but here the whole module fails.
so I'm not sure if it's the same problem as others ;/

wszypelt commented:
@ujfalusi @lgirdwood I'm taking FW and will check these tests on another machine, in a moment I will write how it went.

ujfalusi commented:
> @ujfalusi @lgirdwood I'm taking FW and will check these tests on another machine, in a moment I will write how it went.

Right, so, with this patch the configuration blob is _not_ expected within MOD_INIT_INSTANCE message, a separate LARGE_CONFIG should be use to send the configuration (param_id = 2) and the blob itself should be the exact same payload as it was with IPC3, So in the abi header you need to change the magic to 0x34464F53, set they type to 2 and send that with Linux.

ujfalusi commented:
I was not aware that we had any tests for detect_test at all, like we did not have for the smart_amp_test neither.

mwasko commented:
@iganakov please review proposed changes

wszypelt commented:
@ujfalusi @lgirdwood 
As for the current tests
Unfortunately, all three KD tests fail on this PR. I checked on the same settings, compared it to FW from main. those tests pass there, I also confirmed that the 16bit16bit test is unstable, so I will exclude it from the scope and rerun other PR. The other two KD tests remain.

mszleszy commented:
@lgirdwood 
I'll update tests asap but I need to clarify few things.
KdCfgBlobHeader is being removed from init instance message and it should be send in separate LargeConfigSet.
Should I use IPC4_DETECT_TEST_SET_CONFIG param? 
If so the message will contain base config + KdCfgBlobHeader - do we want to keep it this way, change IPC4_DETECT_TEST_SET_CONFIG  requirements or use different kd param ?

iganakov reviewed (changes requested):
Looks good to me except unused field `load_memory_size` in `sof_detect_test_config`. Probably, should be removed.

aiChaoSONG commented on src/samples/audio/smart_amp_test.c:
```diff
@@ -80,19 +85,31 @@ static struct comp_dev *smart_amp_new(const struct comp_driver *drv,
 #if CONFIG_IPC_MAJOR_4
-	cfg = spec;
-	bs = sizeof(struct sof_smart_amp_config);
+	if (base_cfg->base_cfg_ext.nb_input_pins != 2 ||
+	    base_cfg->base_cfg_ext.nb_output_pins != 1) {
```
We have the macros defined, use them here?
```
#define SMART_AMP_NUM_IN_PINS		2
#define SMART_AMP_NUM_OUT_PINS		1
```

aiChaoSONG commented:
The smart amp change in this PR is verified with #6996 and https://github.com/thesofproject/linux/pull/4160, need more time to test the WoV change.

ujfalusi commented:
I will drop the detect_test change from the PR for now and open a new one for it to get the smart amp feature enabled faster.

ujfalusi commented:
Changes since v1:
- drop the detect_test patch 
- use the defines in place for the number of pins number
- remove duplicated 'ipc4' from the IPC4 config struct name

ujfalusi commented:
> @lgirdwood I'll update tests asap but I need to clarify few things. KdCfgBlobHeader is being removed from init instance message and it should be send in separate LargeConfigSet. Should I use IPC4_DETECT_TEST_SET_CONFIG param? If so the message will contain base config + KdCfgBlobHeader - do we want to keep it this way, change IPC4_DETECT_TEST_SET_CONFIG requirements or use different kd param ?

@mszleszy, the detect_test at init will receive the `struct ipc4_base_module_cfg` only. The configuration is going to arrive via `set_large_config` in a form of (an unmodified, IPC3 compatible) `struct sof_detect_test_config` and the param ID for this config_large is `IPC4_DETECT_TEST_SET_CONFIG` (2).

Fwiw, this is how the blob will look like in topology:
```
CONTROLBYTES_PRIV(DETECTOR_priv,
`       bytes "0x53,0x4f,0x46,0x34,0x02,0x00,0x00,0x00,' # magic + param_id
`       0x20,0x00,0x00,0x00,0x00,0x00,0x00,0x00,'  # size + abi version
`       0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,' # unused
`       0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,' # unused 

`       0x20,0x00,0x00,0x00,0x00,0x00,0x00,0x00,' # the blob itself (struct sof_detect_test_config)
`       0x00,0x00,0x00,0x00,0x34,0x08,0x00,0x00,'
`       0x03,0x00,0x18,0x00,0x00,0x00,0x00,0x00,'
`	0xD0,0x07,0x00,0x00,0x00,0x00,0x00,0x00"'
)

```

ujfalusi commented:
> Looks good to me except unused field `load_memory_size` in `sof_detect_test_config`. Probably, should be removed.

@iganakov, I have intentionally kept the original `struct sof_detect_test_config` intact in order to be able to re-use the same blob from IPC3 in IPC4 with only needing to update the sof_abi_hdr in topology and in user space.
We can not drop members from the struct  from the middle without breaking existing blobs.
